_posts/2016-05-31-unit-vs-integration-akka-testing.md

```diff
@@ -0,0 +1,143 @@
+
+* Check the returned value.
+* Verify calls to collaborators.
+* Inspect the internal state. This might be an smell on some cases.
+
+The concept of returned value is slightly different in Akka. Akka focuses on messages and not on method invocations. Checking the returned value involves two actors and two messages. If the collaborators of the *Actor Under Test* are also Actors, verifying calls will involve two actors and two messages too. If we use a multi-threaded dispatcher for scheduling this, this scenario will be out of our unit testing definition. Let's focus then on testing internal state.
+
+Akka actors are completely encapsulated and the only communication channel is the mailbox. `TestActorRef` is provided by Akka so we can gain access into the internals of an actor so we can unit test it. One of its specialised form is `TestFSMRef` that allows to test [Finite State Machines](http://codurance.com/2016/05/10/finite-state-machines-with-akka/). Let's see an example from our platform:
+
+```scala
+"Item FSM" should {
+  "move into active state when receiving an item" in {
+    val fsm = TestFSMRef(new ItemFSM(itemReportedProducer, itemDeletedBus))
+    fsm.stateName shouldBe Idle
+    fsm.stateData shouldBe Uninitialized
+
+    fsm ! ItemReported(itemId)
+
+    within(200 millis){
+      fsm.stateName shouldBe Active
+      fsm.stateData.asInstanceOf[ItemsToBeDeleted].items shouldBe items
+    }
+  }
+}
+```
+
+As you can see `TestFSMRef` wraps the actor that we want to test and exposes its internal state. That wrapper has other useful methods like setting state programatically or manipulating FSM timers.
+
+I want to share something that confused me a little the first time, but it's important to understand. We need to recall that unit testing in Akka means using one single thread in order to achieve deterministic order of events. `TestActorRef` uses `CallingThreadDispatcher` by default. This dispatcher *runs invocations on the current thread only* so we could 'check the returned value' of an actor with this unit test style.
+
+```scala
+class EchoActor extends Actor {
+ override def receive = {
+   case message ⇒ sender() ! message
+ }
+}
+
+"send back messages unchanged" in {
+  import akka.pattern.ask
+  import scala.concurrent.duration._
+  implicit val timeout = Timeout(5 seconds)
+
+  val actorRef = TestActorRef(new EchoActor)
+
+  val future = actorRef ? "hello world"
+  val Success(result: String) = future.value.get
+  result should be("hello world")
+}
+```
+
+Let's see how we can test this, and other scenarios, with an integration testing style in Akka.
+
+## Akka Integration testing
+
+Akka provides `TestKit` class for integration testing. Let's see one of our tests written using that class:
+
+```scala
+class ItemFSMSpec() extends TestKit(ActorSystem("ItemFSMSpec")) with ImplicitSender
+
+"send complete message to the original sender when one item deleted is received and there are not more pending" in {
```

Comment:
===> send a complete message to the original sender when one deleted item is received and there are no more messages pending"  --- is this what you want to say ?